---
title: "Paper figures (Simulation)"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup dir}
library(ctwas)
library(data.table)
source("~/causalTWAS/causal-TWAS/analysis/summarize_basic_plots.R")
source("~/causalTWAS/causal-TWAS/analysis/summarize_ctwas_plots.R")
source("~/causalTWAS/causal-TWAS/analysis/ld.R")

outputdir = "/home/simingz/causalTWAS/simulations/simulation_ctwas_rss_20210416/"
comparedir = "/home/simingz/causalTWAS/simulations/simulation_ctwas_rss_20210416_compare/"
runtag = "ukb-s80.45-adi"
configtag = 1
pgenfn = "/home/simingz/causalTWAS/ukbiobank/ukb_pgen_s80.45/ukb-s80.45_pgenfs.txt"
ld_pgenfn = "/home/simingz/causalTWAS/ukbiobank/ukb_pgen_s80.45/ukb-s80.45.2_pgenfs.txt"
exprfn = "/home/simingz/causalTWAS/simulations/simulation_ctwas_rss_20210416//ukb-s80.45-adi.expr.txt"
weightf = "/project2/mstephens/causalTWAS/fusion_weights/Adipose_Subcutaneous.pos"

ld_pgenfs <- read.table(ld_pgenfn, header = F, stringsAsFactors = F)[,1]
pgenfs <- read.table(pgenfn, header = F, stringsAsFactors = F)[,1]
pvarfs <- sapply(pgenfs, prep_pvar, outputdir = outputdir)
pgens <- lapply(1:length(pgenfs), function(x) prep_pgen(pgenf = pgenfs[x],pvarf = pvarfs[x]))
exprfs <- read.table(exprfn, header = F, stringsAsFactors = F)[,1]
exprvarfs <- sapply(exprfs, prep_exprvar)

n <- pgenlibr::GetRawSampleCt(pgens[[1]])
p <- sum(unlist(lapply(pgens, pgenlibr::GetVariantCt))) # number of SNPs
J <- 8021 # number of genes

weights <- as.data.frame(fread(weightf, header = T))
weights$ENSEMBL_ID <- sapply(weights$WGT, function(x){unlist(strsplit(unlist(strsplit(x,"/"))[2], "[.]"))[2]})

colorsall <- c("#7fc97f", "#beaed4", "#fdc086")
```

## Parameter estimation
### Parameter vs. true value

In our simulations, the SNP PVE is always set to 0.5 in different settings. The gene PVE is shown as in figures. The number of causal SNPs is always set to 2.5e * $10^{-4}$. In the two settings shown below, number of samples is 45k. For other details about our simulation settings and procedures, please see [here](simulation-ctwas-ukbWG-gtex.adipose_s80.45_041621.html). Note, setting 1 is the high power setting and setting 2 is a low power setting, I will change the setting names manually later. 

Each plot show one parameter: pi.gene, pi.gene/pi.SNP (enrichment), effectsize.gene, PVE.gene. Horizontal bar shows mean true values across the 5 simulations with similar setting parameters. The results by ctwas for each simulation is shown by dots.  

```{r paramplotfunc}
require(latex2exp)
plot_single <- function(mtxlist, truecol, estcol, xlabels = c("setting 1", "setting 2"), ...){
  truth <- do.call(rbind, lapply(1:length(mtxlist), function(x) cbind(x, mean(mtxlist[[x]][, truecol]))))
  est <- do.call(rbind, lapply(1:length(mtxlist), function(x) cbind(x, mtxlist[[x]][, estcol])))
  
  col = est[,1]
  est[,1] <- jitter(est[,1])

  plot(est, pch = 19, xaxt = "n", xlab="" ,frame.plot=FALSE, col = colorsall[col],, las = 1, ...)
  axis(side=1, at=1:2, labels = xlabels, tick = F)
  #text(x=1:length(mtxlist), 0, labels = paste0("temp",1:length(mtxlist)), xpd = T, pos =1)
  for (t in 1:nrow(truth)){
    row <- truth[t,]
    segments(row[1]-0.2, row[2] , row[1] + 0.2, row[2],
       col = colorsall[t], lty = par("lty"), lwd = 2, xpd = FALSE)
  }
  grid()
}
 
get_params <- function(configtag, runtag, simutaglist){
    mtxlist <- list()
  for (group in 1:length(simutaglist)){
    simutags <- simutaglist[[group]]
    source(paste0(outputdir, "config", configtag, ".R"))
    phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
    susieIfs <- paste0(outputdir, runtag, "_simu", simutags, "_config", configtag, ".s2.susieIrssres.Rd")
    susieIfs2 <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".s2.susieIrss.txt")
    mtxlist[[group]] <- show_param(phenofs, susieIfs, susieIfs2, thin = thin)
  }
    return(mtxlist)
}

plot_par <- function(mtxlist){
  par(mfrow=c(1,4))
  plot_single(mtxlist, truecol = "pi1.gene_truth", estcol = "pi1.gene_est", ylab ="gene pi1", ylim = c(0,0.02), xlim = c(0.8,2.4))
  plot_single(mtxlist, truecol = "enrich_truth", estcol = "enrich_est",ylab ="gene enrichment", ylim = c(0,120), xlim = c(0.8,2.4) )
  plot_single(mtxlist, truecol = "sigma.gene_truth", estcol = "sigma.gene_est", ylab = "gene effect size", ylim = c(0.01, 0.03), xlim = c(0.8,2.4))
  plot_single(mtxlist, truecol = "PVE.gene_truth", estcol = "PVE.gene_est", ylab ="gene PVE", ylim = c(0, 0.1), xlim = c(0.8,2.4))
}

```

## Parameters Figure
```{r paramplot, fig.width= 8, fig.height= 3}
simutaglist = list(paste(4, 1:5, sep = "-"), paste(10, 1:5, sep="-"))
mtxlist <- get_params(configtag, runtag, simutaglist)
plot_par(mtxlist)
```

### Supplementary-parameters
```{r paramplot-suppl, fig.width= 8, fig.height= 3}
plot_par_snps <- function(mtxlist){
  par(mfrow=c(1,3))
  plot_single(mtxlist, truecol = "pi1.SNP_truth", estcol = "pi1.SNP_est", xlabels = c("high gene PVE", "low gene PVE"), ylab =TeX('Percent causal, $\\pi_V$'), ylim = c(0,0.0005), xlim = c(0.8,2.4))
  plot_single(mtxlist, truecol = "sigma.SNP_truth", estcol = "sigma.SNP_est",xlabels = c("high gene PVE", "low gene PVE"), ylab = "Variant effect size", ylim = c(0.01, 0.03), xlim = c(0.8,2.4))
  plot_single(mtxlist, truecol = "PVE.SNP_truth", estcol = "PVE.SNP_est",xlabels = c("high gene PVE", "low gene PVE"), ylab ="Variant PVE", ylim = c(0, 0.8), xlim = c(0.8,2.4))
}

plot_par_snps(mtxlist)
```


## PIP calibration
### PIP plot under different settings.

```{r pipplotfunc}
plot_PIP <- function(configtag, runtag,  simutags, ...){
   phenofs <- paste0(outputdir, "ukb-s80.45-adi", "_simu", simutags, "-pheno.Rd")
   susieIfs <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".susieIrss.txt")

   f1 <- caliPIP_plot(phenofs, susieIfs, ...) 
   return(f1)
}
```

```{r pipplot, fig.width= 8, fig.height= 4}
simutaglist = list(paste(4, 1:5, sep = "-"), paste(10, 1:5, sep="-"))
f1 <- plot_PIP(configtag, runtag, simutaglist[[1]], main = "high gene PVE")
f2 <-  plot_PIP(configtag, runtag, simutaglist[[2]], main = "low gene PVE")
gridExtra::grid.arrange(f1, f2, ncol =2)
```

### Supplementary PIP plot

```{r pipplot-suppl, fig.width= 8, fig.height= 16, eval=F}
simutaglist = lapply(c(1:3,5:9), function(x) paste(x, 1:5, sep ="-"))
mtxlist <- get_params(configtag, runtag, simutaglist)
simutaglist = c(list(paste(1, c(1,2,5), sep = "-")), 
                lapply(c(2,3,5,6,7,8,9), function(x) paste(x, 1:5, sep ="-")))

pdf(file = "temp3.pdf", height = 10, width =10)
plotlist <- list()
for (i in 1:length(simutaglist)){
  plotlist[[i]] <- plot_PIP(configtag, runtag, simutaglist[[i]], main = 
                              TeX(sprintf("\\overset{$\\pi_G$= %s, $PVE_G$ = %s;}{$\\pi_V$= %s, $PVE_G$ = %s}",
                                  signif(mean(mtxlist[[i]][,"pi1.gene_truth"]), digits=2), 
                                  signif(mean(mtxlist[[i]][,"PVE.gene_truth"]), digits=2),
                                  signif(mean(mtxlist[[i]][,"pi1.SNP_truth"]), digits=2),
                                  signif(mean(mtxlist[[i]][,"PVE.SNP_truth"]), digits=2))))
}
gridExtra::grid.arrange(plotlist[[1]], plotlist[[2]],
                        plotlist[[3]], plotlist[[4]],
                        plotlist[[5]], plotlist[[6]],
                        plotlist[[7]], plotlist[[8]],ncol = 3)
dev.off()
```

### Comparison with other methods
Bar plot: each bar shows the number of genes, colored by causal status. Use a different color for each method. The method and cut off values:
* ctwas: PIP 0.8
* FUSION fdr: 0.05
* FUSION bonferroni: 0.05
* FUSION permutation: bonferroni < 0.05 then use empirical permutation p < 0.01
* COLOC PP4: 0.8
* FOCUS PIP: 0.8
* SMR FDR: 0.05, updated using ensemble ID
* SMR HEIDI: HEIDI p > 0.05, SMR FDR < 0.05, updated using ensemble ID

Multiple bar plots, different settings: high gene power and low gene power.
```{r ncausalplotfunc}

get_ncausal_df <- function(pfiles, cau, cut = 0.8, useFDP= F,
                           method = c("ctwas", "fusionfdr", "fusionbon","fusionperm", "coloc", "focus", "smr", "smrheidi")){
  df <- NULL
  for (i in 1:length(pfiles)) {
    res <- fread(pfiles[i], header = T)
    # res <- res[complete.cases(res),]
    if (method == "ctwas"){
        res <- data.frame(res[res$type  =="gene", ])
        res$ifcausal <- ifelse(res$id %in% cau[[i]], 1, 0)
        res <- res[order(res$susie_pip, decreasing = T ),] 
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$susie_pip > cut,]}
    } else if (method == "fusionfdr"){
        res$FDR <- p.adjust(res$TWAS.P, method = "fdr")
        res <- res[order(res$FDR, decreasing = F),] 
        res$ifcausal <- ifelse(res$ID %in% cau[[i]], 1, 0)
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$FDR < cut,]}

    } else if  (method == "fusionbon"){
        res$FDR <- p.adjust(res$TWAS.P, method = "bonferroni")
        res <- res[order(res$FDR, decreasing = F),] 
        res$ifcausal <- ifelse(res$ID %in% cau[[i]], 1, 0)
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$FDR < cut,]
        }
    } else if  (method == "fusionperm"){
        res$FDR <- p.adjust(res$TWAS.P, method = "bonferroni")
        res <- res[res$PERM.N >0 & res$PERM.PV < 0.01,]
        res <- res[order(res$FDR, decreasing = F),] 
        res$ifcausal <- ifelse(res$ID %in% cau[[i]], 1, 0)
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$FDR < cut,]
        }
    } else if  (method == "coloc"){
        res <- res[!is.na(res$COLOC.PP4),]
        res <- res[order(res$COLOC.PP4, decreasing = T),] 
        res$ifcausal <- ifelse(res$ID %in% cau[[i]], 1, 0)
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
         res <- res[res$COLOC.PP4 > cut,]
        }
    } else if (method == "focus"){
        res <- res[res$mol_name != "NULL",]
        res$ifcausal <- ifelse(res$mol_name %in% cau[[i]], 1, 0)
        res <- res[order(res$pip, decreasing = T),] 
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$pip > cut, ] 
        }
    }  else if (method == "SMR"){
        res <- as.data.frame(res)
        res$probeID <- sapply(res$Gene, function(x){unlist(strsplit(x, "[.]"))[1]})
        res <- res[res$probeID %in% weights$ENSEMBL_ID,]
        res <- res[sapply(res$p_HEIDI > 0.05, isTRUE),]
        res$FDR <- p.adjust(res$p_SMR, method = "fdr")
        res$ifcausal <- ifelse(res$probeID %in% cau[[i]], 1, 0)
        res <- res[order(res$FDR, decreasing = F),] 
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
           res <- res[res$FDR < cut,]
        }
    } else if (method == "MR-JTI"){
        res$ifcausal <- ifelse(res$variable %in% cau[[i]], 1, 0)
        res <- res[order(abs(res$beta), decreasing = T),] 
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[res$CI_significance=="sig",] 
        }
    } else if (method == "PMR-Egger"){
        res <- as.data.frame(res)
        res$ifcausal <- ifelse(res$gene_id %in% cau[[i]], 1, 0) 
        res <- res[!is.na(res$causal_pvalue), ]
        res <- res[order(res$causal_pvalue, decreasing = F),] 
        res$FDP <- 1-cumsum(res$ifcausal)/(1:nrow(res))
        if (isTRUE(useFDP)){ res <- res[res$FDP < cut,]} else {
          res <- res[sapply(res$causal_pvalue < cut/sum(!is.na(res$causal_pvalue)), isTRUE),]
        }
    } else{
      stop("no such method")
    }

    df.rt <- rbind(c(nrow(res[res$ifcausal == 0, ]), 0, i),
                   c(nrow(res[res$ifcausal == 1, ]), 1, i))
    df <- rbind(df, df.rt)
  }
  colnames(df) <- c("count", "ifcausal", "runtag")
  df <- data.frame(df)
  df$method <- method
  return(df)
}

plot_ncausal <- function(configtag, runtag,  simutags, colors, ...){
  phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
  cau <- lapply(phenofs, function(x) {load(x);get_causal_id(phenores)})
  
  cau_ensembl <- cau
  
  for (i in 1:length(cau_ensembl)){
    cau_ensembl[[i]][cau_ensembl[[i]] %in% weights$ID] <- weights$ENSEMBL_ID[match(cau_ensembl[[i]][cau_ensembl[[i]] %in% weights$ID], weights$ID)]
  }
  
  susieIfs <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".susieIrss.txt")
  fusioncolocfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.coloc.result")
  focusfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.focus.tsv")
  smrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.smr")
  mrjtifs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.mrjti.result")
  pmrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.pmr.result_pi_080")

  ctwas_df <- get_ncausal_df(susieIfs, cau= cau, cut = 0.8, method ="ctwas")
  #fusionfdr_df <- get_ncausal_df(fusioncolocfs, cau= cau, cut = 0.05, method = "fusionfdr")
  fusionbon_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = 0.05, method = "fusionbon")
  fusionperm_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = 0.05, method = "fusionperm",)
  coloc_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = 0.8, method = "coloc")
  focus_df <- get_ncausal_df(focusfs , cau= cau, cut = 0.8, method = "focus")
  #smr_df <- get_ncausal_df(smrfs, cau= cau, cut = 0.05, method = "smr")
  smrheidi_df <- get_ncausal_df(smrfs, cau= cau_ensembl, cut = 0.05, method = "SMR")
  mrjti_df <- get_ncausal_df(mrjtifs, cau= cau_ensembl, method = "MR-JTI")
  pmr_df <- get_ncausal_df(pmrfs, cau= cau_ensembl, cut = 0.05, method = "PMR-Egger")
  
  df <- rbind(ctwas_df, fusionbon_df,fusionperm_df, coloc_df, focus_df, smrheidi_df, mrjti_df, pmr_df)
  df$ifcausal <- df$ifcausal + as.numeric(as.factor(df$method))*10
  df$ifcausal <- as.factor(df$ifcausal)
  fig <- ggbarplot(df, x = "method", y = "count", add = "mean_se", fill = "ifcausal", palette = colors, legend = "none", ...) + grids(linetype = "dashed")
  fig
}
```

```{r ncausalplot, fig.width= 10, fig.height= 4}
colset = c("#ebebeb", "#ffffb3", # FOCUS
           "#ebebeb", "#8dd3c7", # Fusion
           "#ebebeb", "palegreen", # Fusion-permutation
           "#ebebeb", "#CC79A7", # MR-JTI
           "#ebebeb", "goldenrod", #PMR-Egger
           "#ebebeb", "#87CEFA", # SMR
           "#ebebeb", "#fb8072", # cTWAS
           "#ebebeb", "#bebada") # coloc
simutaglist = list(paste(4, 1:5, sep = "-"), paste(10, 1:5, sep="-"))
f1 <- plot_ncausal(configtag, runtag,  simutaglist[[1]], colors = colset, ylim= c(0,225), main = "high gene PVE")
f2 <- plot_ncausal(configtag, runtag,  simutaglist[[2]], colors = colset, ylim= c(0,225), main = "low gene PVE")
gridExtra::grid.arrange(f1, f2, ncol=2)
```

### Comparison with other methods at specified FDP
We can compare the power of methods at a given false discovery proportion, say 10% or 25%. If a method cannot achieve that FDP no matter what the threshold is used, we will point out that fact. 


```{r ncausalplot-FDP, fig.width= 10, fig.height= 4}

plot_ncausal_FDP <- function(configtag, runtag,  simutags, colors, cut =0.2, ...){
  phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
  cau <- lapply(phenofs, function(x) {load(x);get_causal_id(phenores)})
  
  cau_ensembl <- cau
  
  for (i in 1:length(cau_ensembl)){
    cau_ensembl[[i]][cau_ensembl[[i]] %in% weights$ID] <- weights$ENSEMBL_ID[match(cau_ensembl[[i]][cau_ensembl[[i]] %in% weights$ID], weights$ID)]
  }
  
  susieIfs <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".susieIrss.txt")
  fusioncolocfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.coloc.result")
  focusfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.focus.tsv")
  smrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.smr")
  mrjtifs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.mrjti.result")
  pmrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.pmr.result_pi_080")

  ctwas_df <- get_ncausal_df(susieIfs, cau= cau, cut = cut, method ="ctwas", useFDP = T)
  #fusionfdr_df <- get_ncausal_df(fusioncolocfs, cau= cau, cut = 0.05, method = "fusionfdr")
  fusionbon_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = cut, method = "fusionbon", useFDP = T)
  fusionperm_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = cut, method = "fusionperm", useFDP = T)
  coloc_df <- get_ncausal_df(fusioncolocfs , cau= cau, cut = cut, method = "coloc", useFDP = T)
  focus_df <- get_ncausal_df(focusfs , cau= cau, cut = cut, method = "focus", useFDP = T)
  #smr_df <- get_ncausal_df(smrfs, cau= cau, cut = 0.05, method = "smr")
  smrheidi_df <- get_ncausal_df(smrfs, cau= cau_ensembl, cut = cut, method = "SMR", useFDP = T)
  mrjti_df <- get_ncausal_df(mrjtifs, cau= cau_ensembl, cut = cut, method = "MR-JTI", useFDP = T)
  pmr_df <- get_ncausal_df(pmrfs, cau= cau_ensembl, cut = cut, method = "PMR-Egger", useFDP = T)
  
  df <- rbind(ctwas_df, fusionbon_df,fusionperm_df, coloc_df, focus_df, smrheidi_df, mrjti_df, pmr_df)
  df$ifcausal <- df$ifcausal + as.numeric(as.factor(df$method))*10
  df$ifcausal <- as.factor(df$ifcausal)
  fig <- ggbarplot(df, x = "method", y = "count", add = "mean_se", fill = "ifcausal", palette = colors, legend = "none", ...) + grids(linetype = "dashed")
  fig
}

f1 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[1]], colors = colset, cut = 0.2, ylim= c(0,225), main = "high gene PVE")
f2 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[2]], colors = colset, cut = 0.2, ylim= c(0,225), main = "low gene PVE")
gridExtra::grid.arrange(f1, f2, ncol=2)

f1 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[1]], colors = colset, cut = 0.4, ylim= c(0,225), main = "high gene PVE")
f2 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[2]], colors = colset, cut = 0.4, ylim= c(0,225), main = "low gene PVE")
gridExtra::grid.arrange(f1, f2, ncol=2)

f1 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[1]], colors = colset, cut = 1, ylim= c(0,225), main = "high gene PVE")
f2 <- plot_ncausal_FDP(configtag, runtag,  simutaglist[[2]], colors = colset, cut = 1, ylim= c(0,225), main = "low gene PVE")
gridExtra::grid.arrange(f1, f2, ncol=2)
```













## Scatter plot
```{r scatterplot}
runtag <- "ukb-s80.45-adi"
simutags <- paste(3, 2, sep = "-")
phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
gwasfs <- paste0(outputdir, runtag, "_simu",simutags, ".exprgwas.txt.gz")
pipfs <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".susieIrss.txt")
df <- scatter_plot_PIP_p(phenofs, pipfs, gwasfs, main ="PIP-p")
```

## Examples
### FP example 1
cTWAS avoids the FP error. In this case, the false positive gene (from TWAS) is caused by LD of eQTLs with a causal gene's eQTLs.

```{r regionplotfunc}
plot_region <- function(runtag, simutag, configtag, chr, startpos = NULL, endpos = NULL, rerun_ctwas = F, plot_ceqtl = F,
                        plot_eqtl = F){
  pf <- paste0(outputdir, runtag, "_simu",simutag)
  phenof <- paste0(outputdir, runtag, "_simu", simutag, "-pheno.Rd")
  
  b1 <- fread(paste0(pf, ".snpgwas.txt.gz"), header =T)
  setnames(b1, old = "pos", new = "p0")
  b2 <- fread(paste0(pf, ".exprgwas.txt.gz"), header =T)
  b <- rbind(b1, b2, fill = T)
  b <- b[b$chrom == chr,]
  
  susief <- paste0(outputdir, runtag, "_simu", simutag, "_config", configtag, ".susieIrss.txt")
  a <- fread(susief, header =T)
  a <- a[a$chrom == chr,]
  
  if (!is.null(startpos)){
    a <- a[a$pos > startpos & a$pos < endpos]
    b <- b[b$p0 > startpos & b$p0 < endpos]
  }
  
  a <- merge(a, b, by = "id", all = T)
  a[is.na(a$type),"type"] <- "SNP"
  
  if (isTRUE(rerun_ctwas)){
      ld_exprfs <- paste0(outputdir, runtag, "_simu", simutag, "_chr", 1:22, ".expr.gz")
      load(paste0(outputdir, runtag, "_simu", simutag, "_config", configtag, ".s2.susieIrssres.Rd"))
      source(paste0(outputdir, "config", configtag, ".R"))
      group_prior <- group_prior_rec[, ncol(group_prior_rec)] 
      group_prior[2] <- group_prior[2] * thin
      group_prior_var <- group_prior_var_rec[, ncol(group_prior_var_rec)]
      temp_reg <- data.frame("chr" = paste0("chr",chr), "start" = startpos, "stop" = endpos)
      write.table(temp_reg, file= "temp_reg.txt" , row.names=F, col.names=T, sep="\t", quote = F)
      z_gene <-  a[a$type == "gene", c("id", "t.value")]
      colnames(z_gene) <- c("id", "z")
      z_snp <-  a[a$type == "SNP", c("id", "alt", "ref", "t.value")]
      colnames(z_snp) <- c("id", "A1", "A2", "z")
      ctwas_rss(z_gene, z_snp, ld_exprfs= ld_exprfs, ld_pgenfs = ld_pgenfs, ld_R_dir = NULL, ld_regions_custom = "temp_reg.txt", thin = 1, outputdir = ".", outname = "temp", ncore = 1, ncore.rerun = 1, prob_single = 0,  group_prior = group_prior, group_prior_var = group_prior_var, estimate_group_prior = F, estimate_group_prior_var = F)
      a2 <- fread("temp.susieIrss.txt", header = T)
      a <- merge(a2, b, by = "id", all = T)
      # file.remove("temp.susieIrss.txt")
      # file.remove("temp.temp.susieIrssres.Rd")
      # file.remove("temp.regions.txt")
      # file.remove("temp_reg.txt")
   }

  load(phenof)
  cau <- get_causal_id(phenores)
  
  a$ifcausal <- ifelse(a$id %in% cau, 1, 0)
  a[is.na(a$type),"type"] <- "SNP"
  
  a[, "PVALUE"] <- -log10(a[, "PVALUE"])
  a$r2max <- get_ld2(ids =a$id, phenores = phenores, pgenfs = pgenfs, exprfs = exprfs, chrom = chr)
  
  r2cut <- 0.4
  
  layout(matrix(1:2, ncol = 1), widths = 1, heights = c(1.5,1.5), respect = FALSE)
  par(mar = c(0, 4.1, 4.1, 2.1))
  plot(a[a$type=="SNP"]$p0, a[a$type == "SNP"]$PVALUE, pch = 19, xlab="Genomic position" ,frame.plot=FALSE, col = "white", ylim= c(-0.1,1.1), ylab = "ctwas PIP", xaxt = 'n')
  grid()
  points(a[a$type=="SNP"]$p0, a[a$type == "SNP"]$susie_pip, pch = 21, xlab="Genomic position", bg = colorsall[1])
  points(a[a$type=="SNP" & a$r2max > r2cut, ]$p0, a[a$type == "SNP"  & a$r2max >r2cut]$susie_pip, pch = 21, bg = "purple")
  points(a[a$type=="SNP" & a$ifcausal == 1, ]$p0, a[a$type == "SNP" & a$ifcausal == 1]$susie_pip, pch = 21, bg = "salmon")
  points(a[a$type=="gene" ]$p0, a[a$type == "gene" ]$susie_pip, pch = 22, bg = colorsall[1], cex = 2)
  points(a[a$type=="gene" & a$r2max > r2cut, ]$p0, a[a$type == "gene"  & a$r2max > r2cut]$susie_pip, pch = 22, bg = "purple", cex = 2)
  points(a[a$type=="gene" & a$ifcausal == 1, ]$p0, a[a$type == "gene" & a$ifcausal == 1]$susie_pip, pch = 22, bg = "salmon", cex = 2)
  if (isTRUE(plot_ceqtl)){
    gi <- 0
      for (cgene in a[a$type=="gene" & a$ifcausal == 1, ]$id){
     load(paste0(tools::file_path_sans_ext(exprfs[a[a$id == cgene, "chrom.x"][[1]]]), "qc.Rd"))
     eqtls <- rownames(wgtlist[[cgene]])
     points(a[a$id %in% eqtls,]$p0, rep( -0.15 - gi, nrow(a[a$id %in% eqtls,])), pch = "|", col = "salmon", cex = 1.5)
     gi <- gi + 0.2
    }
  }

  if (isTRUE(plot_eqtl)){
     gi <- 0
    for (cgene in a[a$type=="gene" & a$PVALUE >4, ]$id){
     load(paste0(tools::file_path_sans_ext(exprfs[a[a$id == cgene, "chrom.x"][[1]]]), "qc.Rd"))
     eqtls <- rownames(wgtlist[[cgene]])
     points(a[a$id %in% eqtls,]$p0, rep( -0.15-gi, nrow(a[a$id %in% eqtls,])), pch = "|", col = "black", cex = 1.5)
     gi <- gi + 0.2
    }
  }
  
  legend(min(a$p0), y= 1.3 ,c("Gene", "SNP"), pch = c(22,21), title="shape legend", bty ='n', cex =0.8, title.adj = 0)
  legend(min(a$p0), y= 0.8 ,c("Causal", "Noncausal, R2 > 0.4", "Noncausal, R2 <= 0.4"), pch = 19, col = c("salmon", "purple", colorsall[1]), title="color legend", bty ='n', cex =0.8, title.adj = 0)
  par(mar = c(4.1, 4.1, 0.5, 2.1))
  plot(a[a$type=="SNP"]$p0, a[a$type == "SNP"]$PVALUE, pch = 21, xlab="Genomic position" ,frame.plot=FALSE, bg = colorsall[1], ylab = "TWAS -log10(p value)", panel.first = grid(), ylim =c(0, max(a$PVALUE)*1.2))
  points(a[a$type=="SNP" & a$r2max > r2cut ]$p0, a[a$type == "SNP"  & a$r2max > r2cut]$PVALUE, pch = 21, bg = "purple")
  points(a[a$type=="SNP" & a$ifcausal == 1, ]$p0, a[a$type == "SNP" & a$ifcausal == 1]$PVALUE, pch = 21, bg = "salmon")
  points(a[a$type=="gene" ]$p0, a[a$type == "gene" ]$PVALUE, pch = 22, bg = colorsall[1], cex = 2)
  points(a[a$type=="gene" & a$r2max > r2cut, ]$p0, a[a$type == "gene"  & a$r2max > r2cut]$PVALUE, pch = 22, bg = "purple", cex = 2)
  points(a[a$type=="gene" & a$ifcausal == 1, ]$p0, a[a$type == "gene" & a$ifcausal == 1]$PVALUE, pch = 22, bg = "salmon", cex = 2)
  abline(h=-log10(0.05/J), col ="red", lty = 2)
  
  return(a)
}
```

```{r regionplot1, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "4-4"
a <- plot_region(runtag, simutag, configtag, chr = 4, startpos = 43965045, endpos = 45189157,  plot_ceqtl = T, rerun_ctwas = T)
```

### FP example 2
cTWAS avoids the FP error. In this case, the false positive gene (from TWAS) is caused by LD of eQTLs with a causal SNP nearby.

```{r regionplot2, echo=FALSE,results='hide',fig.keep='all', message=F}
runtag = "ukb-s80.45-adi"
simutag <- "3-2"         
a <- plot_region(runtag, simutag, configtag, chr = 19, startpos = 16824416, endpos = 17559862, rerun_ctwas = T, plot_eqtl = T)
```

### TP example
cTWAS is able to find true positives. This gene has one eQTL, cTWAS choose this gene because it uses a  prior favoring genes, it didn't reach significance level by TWAS after bonferron correction.

```{r regionplot3, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "4-4"
a <- plot_region(runtag, simutag, configtag, chr = 1, startpos = 37549183 , endpos =38731847, rerun_ctwas = T)
```


### coloc false positives
```{r regionplot4, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "3-4"
a <- plot_region(runtag, simutag, configtag, chr = 1, startpos = 64000000 , endpos =66000000, rerun_ctwas = F, plot_eqtl = T )
```
```{r regionplot4-1, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "3-4"
a <- plot_region(runtag, simutag, configtag, chr = 1, startpos = 66000000 , endpos =70000000, rerun_ctwas = F, plot_eqtl = T )
```
```{r regionplot4-2, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "3-4"
a <- plot_region(runtag, simutag, configtag, chr = 1, startpos = 155000000 , endpos =170000000, rerun_ctwas = F, plot_eqtl = T )
```
```{r regionplot4-3, echo=FALSE,results='hide',fig.keep='all', message=F}
simutag <- "3-4"
a <- plot_region(runtag, simutag, configtag, chr = 5, startpos = 47000000 , endpos =50000000, rerun_ctwas = F, plot_eqtl = T )
```
