<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>ctwas paper related analyses</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.2.1/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.2.1/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ctwas-paper</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/xinhe-lab/ctwas-paper">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">ctwas paper related analyses</h1>

</div>


<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This repository contains code and data resources to accompany our
research paper:</p>
<blockquote>
<p>Siming, Z., Wesley, C., Sheng, Q., Kaixuan, L., Stephens, M. &amp;
Xin, H. (2022). Adjusting for genetic confounders in transcriptome-wide
association studies leads to reliable detection of causal genes
application to genetic fine mapping. *bioRxiv <a
href="https://doi.org/10.1101/2022.09.27.509700"
class="uri">https://doi.org/10.1101/2022.09.27.509700</a></p>
</blockquote>
<p>The <code>ctwas</code> R package can be accessed <a
href="https://github.com/xinhe-lab/ctwas">here</a>. The main analyses in
the paper was obtained using ctwas v0.1.29.</p>
</div>
<div id="perform-a-test-simulation" class="section level1">
<h1>Perform a test simulation</h1>
<p>An example simulation pipepline is shown as
<code>code/simulation/Snakefile-test</code>. It was written in the <a
href="https://snakemake.readthedocs.io/en/stable/">snakemake</a>
workflow language.</p>
<p>[sheng can you add instructions here]?</p>
</div>
<div id="simulation-performed-in-the-paper." class="section level1">
<h1>Simulation performed in the paper.</h1>
<div id="input-data-for-simulation" class="section level2">
<h2>Input data for simulation</h2>
<ul>
<li>Genotype data.</li>
</ul>
<p>The individual genotype data we used is from UKB biobank, randomly
selecting 80000 or 200000 samples. We then filtered samples. The columns
selected for filtering samples are as follows:</p>
<pre><code>- sex (31)
- UK Biobank assessment centre (54)
- age (21022)
- genetic ethnic grouping (22006)
- genetic sex (22001)
- genotype measurement batch (22000)
- missingness (22005)
- genetic PCs (22009-0.1 - 22009-0.40)
- genetic relatedness pairing (22011)
- genetic kinship (22021)
- outliers (22027)</code></pre>
<p>We filtered the samples based on the following criteria:</p>
<pre><code>- Remove all rows in which one or more of the values are missing, aside from the in the &quot;outlier&quot; and &quot;relatedness_genetic&quot; columns. This step removes any samples that are not marked as being &quot;White British&quot;.
- Remove rows with mismatches between self-reported and genetic sex
- Remove &quot;missingness&quot; and &quot;heterozygosity&quot; outliers as defined by UK Biobank.
- Remove any individuals have at least one relative
- Remove any individuals that have close relatives identified from the &quot;relatendess&quot; calculations that weren&#39;t already identified using the kinship calculations.</code></pre>
<p>This ended up with n = ~45k or ~113k samples used in our simulation.
We use SNP genotype data from chr 1 to chr 22 combined from UKB. We
select SNPs with minor allele frequency &gt; 0.05 and SNPs with at least
95% genotyping rate (5% missing). There are total = 6228664 SNPs passed
these filters and go into our analysis.</p>
<ul>
<li>LD genotype reference</li>
</ul>
<p>We used the genotype of 2k samples from UKbiobank (randomly selected
from the samples used in simulations) to serve as the LD reference.</p>
<ul>
<li>Expression models</li>
</ul>
<p>The one we used in this analysis is GTEx Adipose tissue v7 dataset.
This dataset contains ~ 380 samples, 8021 genes with expression, we used
their lasso results. SNPs included in eQTL anlaysis are restricted to
cis-locus 500kb on either side of the gene boundary. eQTLs are defined
as SNPs with abs(effectize) &gt; 1e-8 in lasso results.</p>
</div>
<div id="running-the-simulation-pipeline" class="section level2">
<h2>Running the simulation pipeline</h2>
<p>Simulation results in our paper comes from four pipelines, available
in the <code>code/simulation/</code> folder.</p>
<p>They are:</p>
<ul>
<li><p><code>Snakefile-simu_20210416</code>, this file simulated GWAS
phenotypes using 45k samples, causal effect size were simulated from
normal distribution.</p></li>
<li><p><code>Snakefile-simu_20210418</code>, this file simulated GWAS
phenotypes using 45k samples, causal effect size were simulated from
normal distribution.</p></li>
<li><p><code>Snakefile-simu_20230321</code>, this file simulated GWAS
phenotypes using 113k samples, causal effect size were simulated from
normal distribution.</p></li>
</ul>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
